File: design-patterns-ou-padroes-de-projeto/behavioral-patterns/unit-of-work/unit-of-work.py
class UnitOfWork:
    def __init__(self):
        self._operations = []

    def register(self, operation):
        self._operations.append(operation)

    def commit(self):
        for operation in self._operations:
            operation()

# Exemplo real de uso
class UserRegistrationService:
    @staticmethod
    def call(user_params, profile_params):
        uow = UnitOfWork()

        user = User(user_params)
        profile = Profile(profile_params)

        uow.register(lambda: user.save())
        uow.register(lambda: (setattr(profile, 'user', user), profile.save()))
        uow.register(lambda: AuditLog.create("UserCreated", user.to_json()))

        uow.commit()

# Mock de classes auxiliares
class User:
    def __init__(self, params):
        pass
    def save(self):
        pass
    def to_json(self):
        return "{}"

class Profile:
    def __init__(self, params):
        self.user = None
    def save(self):
        pass

class AuditLog:
    @staticmethod
    def create(action, data):
        pass

File: design-patterns-ou-padroes-de-projeto/behavioral-patterns/unit-of-work/unit-of-work.cs
using System;
using System.Collections.Generic;

public class UnitOfWork
{
    private readonly List<Action> _operations = new List<Action>();

    public void Register(Action operation)
    {
        _operations.Add(operation);
    }

    public void Commit()
    {
        foreach (var operation in _operations)
        {
            operation();
        }
    }
}

// Exemplo real de uso
public class UserRegistrationService
{
    public static void Call(Dictionary<string, object> userParams, Dictionary<string, object> profileParams)
    {
        var uow = new UnitOfWork();

        var user = new User(userParams);
        var profile = new Profile(profileParams);

        uow.Register(() => user.Save());
        uow.Register(() => { profile.User = user; profile.Save(); });
        uow.Register(() => AuditLog.Create("UserCreated", user.ToJson()));

        uow.Commit();
    }
}

// Mock de classes auxiliares
public class User
{
    public User(Dictionary<string, object> parameters) { /* atribuição */ }
    public void Save() { /* salvar no DB */ }
    public string ToJson() => "{}";
}

public class Profile
{
    public User User { get; set; }
    public Profile(Dictionary<string, object> parameters) { }
    public void Save() { }
}

public static class AuditLog
{
    public static void Create(string action, string data) { }
}
